### PR TITLE
Create a GKE deployer for kubetest2.

### PR DESCRIPTION
**Pull request by RonWeber**
Works with --up and --down, probably not --test

### CODE REVIEW

File: kubetest2/pkg/exec/exec.go [hunk 1]
```diff
@@ -71,3 +93,10 @@ func InheritOutput(cmd Cmd) {
 	cmd.SetStderr(os.Stderr)
 	cmd.SetStdout(os.Stdout)
 }
+
+// NoOutput ignores all output from the command.
+func NoOutput(cmd Cmd) {
+	var void bytes.Buffer
+	cmd.SetStdout(&void)
```
Comment (BenTheElder) [inline kubetest2/pkg/exec/exec.go]: use ioutil.Discard
https://golang.org/pkg/io/ioutil/#pkg-variables

File: kubetest2/kubetest2-gke/deployer/deployer.go [hunk 2]
```diff
@@ -0,0 +1,683 @@
...
+		if len(numNodeLocations) > 1 {
+			if err := os.Setenv("MULTIZONE", "true"); err != nil {
+				return fmt.Errorf("error setting MULTIZONE env variable: %v", err)
+			}
+		}
+	}
+
+	fmt.Printf("Environment: %v", os.Environ())
+
+	// TODO(zmerlynn): The version should be plumbed through Extract
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/deployer.go]: this new TODO should probably be you :-)

File: kubetest2/kubetest2-gke/deployer/deployer.go [hunk 3]
```diff
@@ -0,0 +1,683 @@
...
+			"--project=" + d.project,
+			loc,
+			"--machine-type=" + pool.MachineType,
+			"--num-nodes=" + strconv.Itoa(pool.Nodes)}
+		poolArgs = append(poolArgs, pool.ExtraArgs...)
+		if err := runWithOutput(exec.Command("gcloud", d.containerArgs(poolArgs...)...)); err != nil {
+			return fmt.Errorf("error creating node pool %q: %v", poolName, err)
+		}
+	}
+	log.Printf("Line 336")
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/deployer.go]: stray development debugging?

File: kubetest2/kubetest2-gke/deployer/deployer.go [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+		return false, metadata.NewJUnitError(err, strings.Join(lines, "\n"))
+	}
+	return len(lines) > 0, nil
+}
+
+// DumpClusterLogs for GKE generates a small script that wraps
+// log-dump.sh with the appropriate shell-fu to get the cluster
+// dumped.
+//
+// TODO(zmerlynn): This whole path is really gross, but this seemed
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/deployer.go]: also probably should be you :-)

File: kubetest2/pkg/exec/exec.go (hunk 1 above)
Comment (RonWeber) [reply kubetest2/pkg/exec/exec.go]: Done.

File: kubetest2/kubetest2-gke/deployer/deployer.go (hunk 2 above)
Comment (RonWeber) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: Fair enough.

File: kubetest2/kubetest2-gke/deployer/deployer.go (hunk 3 above)
Comment (RonWeber) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: Thought I got all those...

File: kubetest2/kubetest2-gke/deployer/deployer.go (hunk 4 above)
Comment (RonWeber) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: Yeah

File: kubetest2/kubetest2-gke/deployer/commandutils.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"k8s.io/test-infra/kubetest2/pkg/exec"
+)
+
+func (d *deployer) prepareGcpIfNeeded() error {
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/commandutils.go]: this code could do with some more refactoring and documenting, but I understand that this is mostly a transplant of functionality reverse engineered from kubetest :-)

File: kubetest2/kubetest2-gke/deployer/commandutils.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+func activateServiceAccount(path string) error {
+	if path == "" {
+		return nil
+	}
+	return runWithOutput(exec.Command("gcloud", "auth", "activate-service-account", "--key-file="+path))
+}
+
+// home returns $HOME/part/part/part
+func home(parts ...string) string {
+	p := []string{os.Getenv("HOME")}
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/commandutils.go]: this doesn't look right

Comment (BenTheElder) [reply kubetest2/kubetest2-gke/deployer/commandutils.go]: I think this should just be filepath.Join?

File: kubetest2/kubetest2-gke/deployer/commandutils.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	for _, a := range parts {
+		p = append(p, a)
+	}
+	return filepath.Join(p...)
+}
+
+// pushEnv pushes env=value and return a function that resets env
+func pushEnv(env, value string) (func() error, error) {
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/commandutils.go]: we should probably refactor this out to just pass it to the target process, popping env is a bit of an antipattern

File: kubetest2/kubetest2-gke/deployer/deployer.go [hunk 8]
```diff
@@ -0,0 +1,681 @@
...
+		return err
+	}
+	var filters []string
+	for _, ig := range d.instanceGroups {
+		filters = append(filters, fmt.Sprintf("(metadata.created-by:*%s)", ig.path))
+	}
+
+	// Generate the log-dump.sh command-line
+	var dumpCmd string
+	if d.gcsLogsDir == "" {
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/deployer.go]: seems this can just be the first case, and then optionally append the second case?

Comment (RonWeber) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: Done.

File: kubetest2/kubetest2-gke/deployer/commandutils.go (hunk 6 above)
Comment (RonWeber) [reply kubetest2/kubetest2-gke/deployer/commandutils.go]: It's not entirely wrong, just ugly.  Changed.

Comment (BenTheElder) [reply kubetest2/kubetest2-gke/deployer/commandutils.go]: do you mind also updating the doc comment? I also don't see a change?

Comment (BenTheElder) [reply kubetest2/kubetest2-gke/deployer/commandutils.go]: /part/part/part is not the most useful or correct comment, E.G. it's doing filepath.Join which may use the host platform's seperator. /shrug

File: kubetest2/kubetest2-gke/deployer/commandutils.go (hunk 7 above)
Comment (RonWeber) [reply kubetest2/kubetest2-gke/deployer/commandutils.go]: Done.

File: kubetest2/kubetest2-gke/deployer/deployer.go [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+		errFirewall = exec.Command("gcloud", "compute", "firewall-rules", "delete", "-q", firewall,
+			"--project="+d.project).Run()
+	} else {
+		log.Printf("Found no rules for firewall '%s', assuming resources are clean", firewall)
+	}
+	numLeakedFWRules, errCleanFirewalls := d.cleanupNetworkFirewalls()
+	var errSubnet error
+	errNetwork := runWithOutput(exec.Command("gcloud", "compute", "networks", "delete", "-q", d.network,
+		"--project="+d.project))
+	if errCluster != nil {
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/deployer.go]: we should consider using a multi-error type to collect up and return all of these in the future

File: kubetest2/kubetest2-gke/deployer/deployer.go [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+	d.testPrepared = true
+	return nil
+}
+
+func (d *deployer) getKubeConfig() error {
+	info, err := os.Stat(d.kubecfg)
+	if err != nil {
+		return err
+	}
+	if info.Size() > 0 {
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/deployer.go]: this check seems strange

File: kubetest2/kubetest2-gke/deployer/deployer.go [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+
+	// don't delete default network
+	if d.network == "default" {
+		if errCluster != nil {
+			log.Printf("Error deleting cluster using default network, allow the error for now %s", errCluster)
+		}
+		return nil
+	}
+
+	var errFirewall error
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/deployer.go]: we should consider breaking some of these out into higher level subroutines in the future like "delete firewalls"

File: kubetest2/kubetest2-gke/deployer/deployer.go [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+	if err != nil {
+		return fmt.Errorf("instance group URL fetch failed: %s", execError(err))
+	}
+	igURLs := strings.Split(strings.TrimSpace(string(igs)), ";")
+	if len(igURLs) == 0 {
+		return fmt.Errorf("no instance group URLs returned by gcloud, output %q", string(igs))
+	}
+	sort.Strings(igURLs)
+	for _, igURL := range igURLs {
+		m := poolRe.FindStringSubmatch(igURL)
```
Comment (BenTheElder) [inline kubetest2/kubetest2-gke/deployer/deployer.go]: fantastic

File: kubetest2/kubetest2-gke/deployer/commandutils.go (hunk 5 above)
Comment (BenTheElder) [reply kubetest2/kubetest2-gke/deployer/commandutils.go]: this is a lot better now

File: kubetest2/kubetest2-gke/deployer/deployer.go (hunk 9 above)
Comment (BenTheElder) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: https://github.com/kubernetes/apimachinery/blob/b74e4c97951f106696db86d7593effcb1463ca30/pkg/util/errors/errors.go#L31

Comment (BenTheElder) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: consider using a type like that (k8s.io/apimachinery/pkg/util/errors.Aggregate)

Comment (BenTheElder) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: (can be done in a follow-up)

File: kubetest2/kubetest2-gke/deployer/deployer.go (hunk 10 above)
Comment (RonWeber) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: There's an almost-logical explanation: Somewhere far away, we use ioutil.TempFile to set d.kubecfg to a filename we can safely overwrite.  But ioutil.TempFile creates the file.  So we leave it empty and make our real existence check "size > 0".

Comment (BenTheElder) [reply kubetest2/kubetest2-gke/deployer/deployer.go]: we should probably refactor that to a tempdir :-)

**Pull request merged**